R/generate_data.R

```diff
@@ -385,38 +392,68 @@ genFormula <- function(coefs, vars) {
 #' @concept generate_data
 genMarkov <- function(n, transMat, chainLen, wide = FALSE, id = "id",
                       pername = "period", varname = "state",
-                      widePrefix = "S", trimvalue = NULL) {
+                      widePrefix = "S", trimvalue = NULL, startProb = NULL) {
 
   # 'declare' vars created in data.table
   variable <- NULL
 
-  # check transMat is square matrix and row sums = 1
-
-  if (!is.matrix(transMat) |
-    (length(dim(transMat)) != 2) |
-    (dim(transMat)[1] != dim(transMat)[2])
-  ) {
-    stop("Transition matrix needs to be a square matrix")
+  # check transMat is matrix
+  #assertType(var1 = transMat, type = "matrix", deep = FALSE)
+  if (!is.matrix(transMat)) {
+    c <- condition(c("simstudy::typeMatrix", "error"),
+                   "transMat is not a matrix!")
+    stop(c)
   }
-
-  # check row sums = 1
-
+    
+  # check transMat is square matrix
+  if ((length(dim(transMat)) != 2) |
+      (dim(transMat)[1] != dim(transMat)[2])) {
+    c <- condition(c("simstudy::squareMatrix", "error"),
+                   "transMat is not a square matrix!")
+    stop(c)
+  }
+  
+  # check transMat row sums = 1
   if (!all(round(apply(transMat, 1, sum), 5) == 1)) {
-    stop("Rows in transition matrix must sum to 1")
+    c <- condition(c("simstudy::rowSums1", "error"),
+                   "transMat rows do not sum to 1!")
+    stop(c)
+  }
+  
+  # check chainLen greater than 1
+  if (chainLen <= 1) {
+    c <- condition(c("simstudy::chainLen", "error"),
+                   "chainLen must be greater than 1!")
+    stop(c)
+  }
+  
+  # if startProb defined, check it has length == number of matrix rows
+  if (!is.null(startProb)) {
+    s <- unlist(strsplit(startProb, split = ";"))
+    r <- dim(transMat)[1]
+      assertLength(var1 = s, length = r)
+    
   }
-
-  # check chainLen is > 1
-
-  if (chainLen <= 1) stop("Chain length must be greater than 1")
 
   ####
 
-  dd <- genData(n = n, id = id)
-  dd <- addMarkov(dd, transMat, chainLen, wide, id,
-    pername, varname, widePrefix,
-    start0lab = NULL,
-    trimvalue = trimvalue
-  )
+  if (!is.null(startProb)) {
+    dprob <- defData(varname = "prob", formula = startProb, dist = "categorical")
+    dd <- genData(n = n, dprob, id = id)
+    dd <- addMarkov(dd, transMat, chainLen, wide, id,
+      pername, varname, widePrefix,
+      start0lab = "prob",
+      trimvalue = trimvalue
+    )
```

Comment:
This seems to work quite well, but I when the `startProb` probabilities do not add up to 1, it can generate an error.

```
library(simstudy)

P <- matrix(c(.4, .5, .1,
              .1, .3, .6,
              .3, .4, .3), 
  nrow = 3, byrow = T)

genMarkov(101, P, 2, wide = TRUE, startProb = "0.3;0.3;0.1")
#> Warning: Probabilities do not sum to 1. Adding category to all rows!
#> Error in markovChains(n, transMat, chainLen, s0): Row index is out of bounds: [index=3; row extent=3].
```

`genData` is written so that If you generate a categorical variable with probabilities that add up to less than 1, another category is added (and a warning will be generated). So, in this case, if we specify `"0.3;0.3;0.1", a fourth category with probability .3 will be added:

```
d2 <- defData(varname = "c", formula = "0.3;0.3;0.1", dist = "categorical")
genData(10000, d2)[, table(c)]
#> Warning: Probabilities do not sum to 1. Adding category to all rows!
#> c
#>    1    2    3    4 
#> 3057 2975  986 2982
```

This creates problems for `genMarkov` because there are now 4 start states and the transition matrix only has 3 states - and that is why the error is generated by `genMarkov`. The easiest solution is probably to a specific check in `genMarkov` to make sure the probabilities sum to 1. But, I am open to other solutions.

And this brings to mind another possible error with `addMarkov` - if the user generates start states that don't exist in the transition matrix, errors could be generated. Could you check that out as well?

Just as an aside, if probabilities add up to more than 1, the function normalizes the probabilities to ensure that the probabilities do add up to 1, generates the data, and issues a warning. This does not lead to any problems for `genMarkov`, because there are the correct number of categories.

Reply:
I believe I addressed both of these things, take a look

Reply:
It definitely looks like you have implemented the check - so that it does indeed issue a warning when the startProb sum is not 1. But when `startProb` is specified, `genMarkov` fails (and your test picks that up). So - this generates an error:

```
genMarkov(101, P, 2, wide = TRUE, startProb = "0.5;0.3;0.2")
```

But this does not:

```
genMarkov(101, P, 2, wide = TRUE)
```

Reply:
Yep I just caught that and updated it